metricbeat/internal/metrics/memory/memory.go

```diff
@@ -0,0 +1,102 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package memory
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/internal/metrics"
+)
+
+// Memory holds os-specifc memory usage data
+// The vast majority of these values are cross-platform
+// However, we're wrapping all them for the sake of safety, and for the more variable swap metrics
+type Memory struct {
+	Total metrics.OptUint `struct:"total,omitempty"`
+	Used  UsedMemStats    `struct:"used,omitempty"`
+
+	Free   metrics.OptUint `struct:"free,omitempty"`
+	Cached metrics.OptUint `struct:"cached,omitempty"`
+	// "Actual" values are, technically, a linux-only concept
+	// For better or worse we've expanded it to include "derived"
+	// Memory values on other platforms, which we should
+	// probably keep for the sake of backwards compatibility
+	// However, because the derived value varies from platform to platform,
+	// We may want to more precisely document what these mean.
+	Actual ActualMemoryMetrics `struct:"actual,omitempty"`
+
+	// Swap metrics
+	Swap SwapMetrics `struct:"swap,omitempty"`
+}
+
+// UsedMemStats wraps used.* memory metrics
+type UsedMemStats struct {
+	Pct   metrics.OptFloat `struct:"pct,omitempty"`
+	Bytes metrics.OptUint  `struct:"bytes,omitempty"`
+}
+
+// ActualMemoryMetrics wraps the actual.* memory metrics
+type ActualMemoryMetrics struct {
+	Free metrics.OptUint `struct:"free,omitempty"`
+	Used UsedMemStats    `struct:"used,omitempty"`
+}
+
+// SwapMetrics wraps swap.* memory metrics
+type SwapMetrics struct {
+	Total metrics.OptUint `struct:"total,omitempty"`
+	Used  UsedMemStats    `struct:"used,omitempty"`
+	Free  metrics.OptUint `struct:"free,omitempty"`
+}
+
+// Get returns platform-independent memory metrics.
+func Get(procfs string) (Memory, error) {
+	base, err := get(procfs)
+	if err != nil {
+		return Memory{}, errors.Wrap(err, "error getting system memory info")
+	}
+	base.fillPercentages()
+	return base, nil
+}
+
+// IsZero implements the zeroer interface for structform's folders
+func (used UsedMemStats) IsZero() bool {
+	return used.Pct.IsZero() && used.Bytes.IsZero()
+}
+
+// IsZero implements the zeroer interface for structform's folders
+func (swap SwapMetrics) IsZero() bool {
+	return swap.Free.IsZero() && swap.Used.IsZero() && swap.Total.IsZero()
+}
+
+func (base *Memory) fillPercentages() {
+	// Add percentages
+	// In theory, `Used` and `Total` are available everywhere, so assume values are good.
+	if base.Total.ValueOr(0) != 0 {
+		percUsed := float64(base.Used.Bytes.ValueOr(0)) / float64(base.Total.ValueOr(1))
```

Comment:
what is percUsed to be if `base.Used.Bytes.IsZero()` returns true? We can't compute a percentage for someting that does not exist.

As I already mentioned a helper for Add/Sub, maybe it's worth to have one for div?
The computation for percentages is done on floats. Maybe we also need a 'cast' helper?

Assuming we have helpers for math, the calls would become:


```
base.Used.Pct = based.Used.Bytes.AsFloat().Div(base.Total.AsFloat().ValueOr(1)).Round(common.DefaultDecimalPlacesCount)
base.Actual.Used.Pct = ...
```



An 'abstraction' to Add/Sub and co would be (now we are moving into FP land... I wish we had generics, then we could go from `Opt[uint] -> Opt[float64]`):

```
func (opt OptUint) Update(fn func(uint64) uint64)) (out OptUint) {
  if opt.exists {
    out = OptUintWith(fn(opt.value))
  }
  return out
}
```

In general: You should not compute percentages or other derived values if the original value is not set.

Reply:
> (now we are moving into FP land... I wish we had generics, then we could go from Opt[uint] -> Opt[float64])

Yah, a common refrain of mine throughout this entire refactor.

Also, didn't think of that, I was mostly worried about making sure the denominator wasn't zero. I'm a little hesitant to build up long chains of `.thing().thing().thing()`, since those can be a tad difficult to parse. Our usage of `Round(common.DefaultDecimalPlacesCount)` and `float64` is pretty consistent throughout the codebase, so we might either want to just implicitly code that logic in, and/or provide another `NewX` function that allows us to override it. That way, `.Div()` would look a lot less ugly in practice.

Reply:
Yeah. For values that represent percentages, we might use the rounding when reporting the value.

> is pretty consistent throughout the codebase, so we might either want to just implicitly code that logic in, and/or provide another NewX function

Rounding adds an error. Computing based on rounded values and rounding again increases that error. That is, I would be in favor of having a rounding function or method that must be used explicitely, or do the rounding only when reporting the value.

> I'm a little hesitant to build up long chains

Yeah, these can become hard to read and should not be encouraged to use. But then, one can always create temporary variables. The way you solved it with the tmp + round here looks good. But now we always require people to remember: Only compute a derived value if the opt is actually set. One can not derive a value from None.